File: trunk/projects/CAndCPPCodec/src/RL_glue.c
/* 
* Copyright (C) 2007, Adam White

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#include <RL_glue.h>

static Action last_action     = {0};
static Observation last_state = {0};

static Reward last_reward  = 0;
static Reward total_reward = 0;

static int is_terminal     = 0;
static int num_steps       = 0;
static int total_steps     = 0;
static int num_episodes    = 0;

static Reward_observation_action_terminal roa;
static Observation_action oa;

void RL_init() {
  Task_specification task_spec = 0;

  task_spec = env_init();
  agent_init(task_spec);

  last_reward     = 0;
  total_reward    = 0;

  num_steps       = 0;
  is_terminal     = 0;
  total_steps     = 0;
  num_episodes    = 0;
}

Observation_action RL_start() {
  num_steps    = 1;
  is_terminal  = 0;
  total_reward = 0;

  last_state = env_start();
  last_action = agent_start(last_state);
  
  oa.o = last_state;
  oa.a = last_action;
  
  return oa;
}

Reward_observation_action_terminal RL_step() {
  Reward_observation ro;
  ro = env_step(last_action);
  last_reward = ro.r;
  last_state = ro.o;
  
  roa.r = ro.r;
  roa.o = ro.o;
  roa.terminal = ro.terminal;
  
  is_terminal = ro.terminal;
  total_reward += last_reward;

  if (ro.terminal == 1) {
    num_episodes += 1;
    total_steps  += num_steps;
    agent_end(last_reward);
  }
  else {
    last_action = agent_step(last_reward,last_state);
    num_steps += 1; /* increment num_steps if state is not terminal */
    roa.a = last_action;
  }
  return roa;
}

void RL_cleanup() {
  env_cleanup();
  agent_cleanup();
}

Message RL_agent_message(const Message message) {
  return agent_message(message);
}

Message RL_env_message(const Message message) {
  return env_message(message);
}

void RL_freeze() {
  agent_freeze();
}

void RL_episode(unsigned int maxStepsThisEpisode) {
  unsigned int x = 0;
  RL_start();
/* RL_start sets current step to 1, so we should start x at 1 */
  for ( x = 1; !is_terminal && (maxStepsThisEpisode == 0 ? 1 : x < maxStepsThisEpisode); ++x ) {
    RL_step();
  }
}

Reward RL_return() {
  return total_reward;
}

int RL_num_steps() {
  /* number of steps of the current or just completed episodes of run */
  return num_steps;
}

int RL_num_episodes() {
  return num_episodes;
}

State_key RL_get_state() {
  return env_get_state();
}

void RL_set_state(State_key sk) {
  env_set_state(sk);
}

void RL_set_random_seed(Random_seed_key rsk) {
  env_set_random_seed(rsk);
}

Random_seed_key RL_get_random_seed() {
  return env_get_random_seed();
}

File: projects/codecs/Python/src/rlglue/RLGlue.py
# 
# Copyright (C) 2007, Mark Lee
# 
#http://rl-glue.googlecode.com/
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys

import rlglue.network.Network as Network
from rlglue.types import Observation_action
from rlglue.types import Random_seed_key 
from rlglue.types import Reward_observation_action_terminal
from rlglue.types import State_key

network = None

# () -> void
def forceConnection():
	global network
	if network == None:
		network = Network.Network()
		network.connect()
		network.clearSendBuffer()
		network.putInt(Network.kExperimentConnection)
		network.putInt(0)
		network.send()

# (int) -> void
def doStandardRecv(state):
	network.clearRecvBuffer()
	recvSize = network.recv(8) - 8
	
	glueState = network.getInt()
	dataSize = network.getInt()
	remaining = dataSize - recvSize
	
	if remaining < 0:
		remaining = 0
	
	remainingReceived = network.recv(remaining)
	
	# Already read the header, so discard it
	network.getInt()
	network.getInt()
	
	if (glueState != state):
		sys.stderr.write("Not synched with server. glueState = " + str(glueState) + " but should be " + str(state) + "\n")
		sys.exit(1)

# (int) -> void
def doCallWithNoParams(state):
	network.clearSendBuffer()
	network.putInt(state)
	network.putInt(0)
	network.send()

#Brian Tanner... need to make this return a string
# () -> string
def RL_init():
	forceConnection()
	doCallWithNoParams(Network.kRLInit)
	doStandardRecv(Network.kRLInit)
	#Brian Tanner added
	taskSpecResponse = network.getString()
	return taskSpecResponse

# () -> Observation_action
def RL_start():
	obsact = None
	doCallWithNoParams(Network.kRLStart)
	doStandardRecv(Network.kRLStart)
	obsact = Observation_action()
	obsact.o = network.getObservation()
	obsact.a = network.getAction()
	return obsact

# () -> Reward_observation_action_terminal
def RL_step():
	roat = None
	doCallWithNoParams(Network.kRLStep)
	doStandardRecv(Network.kRLStep)
	roat = Reward_observation_action_terminal()
	roat.terminal = network.getInt()
	roat.r = network.getDouble()
	roat.o = network.getObservation()
	roat.a = network.getAction()
	return roat

# () -> void
def RL_cleanup():
	doCallWithNoParams(Network.kRLCleanup)
	doStandardRecv(Network.kRLCleanup)

# (string) -> string
def RL_agent_message(message):
	if message == None:
		message=""
	response = ""
	forceConnection()
	network.clearSendBuffer()
	network.putInt(Network.kRLAgentMessage)
	#Payload Size
	network.putInt(len(message) + 4)
	network.putString(message)
	network.send()
	doStandardRecv(Network.kRLAgentMessage)
	response = network.getString()
	return response

# (string) -> string
def RL_env_message(message):
	if message == None:
		message=""
	response = ""
	forceConnection()
	network.clearSendBuffer()
	network.putInt(Network.kRLEnvMessage)
	#Payload Size
	network.putInt(len(message) + 4)
	network.putString(message)
	network.send()
	doStandardRecv(Network.kRLEnvMessage)
	response = network.getString()
	return response

# () -> double
def RL_return():
	reward = 0.0
	doCallWithNoParams(Network.kRLReturn)
	doStandardRecv(Network.kRLReturn)
	reward = network.getDouble()
	return reward

# () -> int
def RL_num_steps():
	numSteps = 0
	doCallWithNoParams(Network.kRLNumSteps)
	doStandardRecv(Network.kRLNumSteps)
	numSteps = network.getInt()
	return numSteps

# () -> int
def RL_num_episodes():
	numEpisodes = 0
	doCallWithNoParams(Network.kRLNumEpisodes)
	doStandardRecv(Network.kRLNumEpisodes)
	numEpisodes = network.getInt()
	return numEpisodes

#Brian Tanner needs to make this return an int
# (int) -> int
def RL_episode(num_steps):
	network.clearSendBuffer()
	network.putInt(Network.kRLEpisode)
	network.putInt(Network.kIntSize)
	network.putInt(num_steps)
	network.send()
	doStandardRecv(Network.kRLEpisode)
	#Brian Tanner added
	exitStatus = network.getInt()
	return exitStatus

#This is deprecated, should remove it
# () -> void
def RL_freeze():
	doCallWithNoParams(Network.kRLFreeze)
	doStandardRecv(Network.kRLFreeze)

# (State_key) -> void
def RL_set_state(sk):
	if sk == None:
		sk=State_key()
	network.clearSendBuffer()
	network.putInt(Network.kRLSetState)
	network.putInt(network.sizeOfStateKey(sk))
	network.putStateKey(sk)
	network.send()
	doStandardRecv(Network.kRLSetState)

# (Random_seed_key) -> void
def RL_set_random_seed(rsk):
	if rsk == None:
		rsk=Random_seed_key()
	network.clearSendBuffer()
	network.putInt(Network.kRLSetRandomSeed)
	network.putInt(network.sizeOfRandomSeed(rsk))
	network.putRandomSeedKey(rsk)
	network.send()
	doStandardRecv(Network.kRLSetRandomSeed)

# () -> State_key
def RL_get_state():
	key = None
	doCallWithNoParams(Network.kRLGetState)
	doStandardRecv(Network.kRLGetState)
	key = network.getStateKey()
	if key == None:
		key=State_key()
	return key

# () -> Random_seed_key
def RL_get_random_seed():
	key = None
	doCallWithNoParams(Network.kRLGetRandomSeed)
	doStandardRecv(Network.kRLGetRandomSeed)
	key = network.getRandomSeedKey()
	if key == None:
		key=Random_seed_key()
	return key